types/react/test/hooks.tsx

```diff
@@ -270,18 +276,31 @@ function useEveryHook(ref: React.Ref<{ id: number }> | undefined): () => boolean
     // make sure setState accepts a function
     setToggle(r => !r);
 
-    // Undesired
-    // Should not type-check since `number` will be `number` at runtime but `() => number` in the type-checker
-    const [number, setNumber] = React.useState<() => number>(() => 0);
-    // Should be `number`
-    // $EpectType () => number
+    // $ExpectType number
+    React.useState(() => 0)[0];
+    // $ExpectType string
+    React.useState(() => "")[0];
+
+    React.useState();
+
+    // make sure optional functions signatures works as expected
+    const [number, setNumber] = React.useState<() => number>();
+    // $ExpectType (() => number) | undefined
     number;
+    setNumber();
+    // When storing a function, don't allow direct input to update the state
+    // @ts-expect-error
+    setNumber(() => 0);
+    setNumber(() => () => 0);
+    // When storing a function, don't allow direct input to initialize the state
+    // @ts-expect-error
+    React.useState<() => number>(() => 0);
 
     const [numFunc, setNumFunc] = React.useState<() => number>(() => () => 0);
     // $ExpectType () => number
     numFunc;
-    // Undesired
-    // Should not typecheck since that would update the state to `number` when the type-checker would still consider the state to be `() => number`
+    // When storing a function, don't allow direct input to update the state
+    // @ts-expect-error
```

Comment:
This is how every changed test should be structured. We have a direct mapping between undesired behavior and a `@ts-expect-error`